### ISSUE TITLE
Better console output

### ISSUE THREAD

**Issue by vlascik** [ethereum-mining/ethminer#62]
Console output would be easier to grasp if:
- console averaged speed output over several (3, 5, or configurable) seconds, so that the rate doesn't jump up and down and scroll down too quickly
- speed was printed out on card by card basis too, so that it would be easier to find out which card could be improved on
- number of shares and number of hours since the start was in the output too
- some ansi colors would be used

**Comment by renaudcerrato** [comment]
Absolutely need a better console output for automated monitoring (at least). I'm used to write wrappers in python to add some kind of advanced "watchdog" feature and/or integration with systemd. The current output is incomplete enough to be usefull.

**Comment by piotr-dobrogost** [comment]
@renaudcerrato 
As you mentioned systemd, can you reference any systemd service files for controlling ethminer?

**Comment by JHGitty** [comment]
It would be nice to have terminal color support on Windows. 👍

**Comment by oleng** [comment]
to add the sentiment, bottom line is: miner _loves_ stats & control.

look how popular claymore is, it's because he gave them what they want (and need)

**Comment by bsodmike** [comment]
It would be nice if shares are given a colour - Yellow or bright Green in the console output when a share is found and accepted.

Thanks!

**Comment by evilny0** [comment]
See https://github.com/ethereum-mining/ethminer/pull/164 . I added colors for accepted & refused shares, and hashrate per card.

**Comment by Jacxz** [comment]
Uptime was added with #306 
Hashrate per card was added with #225

**Comment by 4a48** [comment]
As more and more GPUs are supported per rig, it becomes that much more important to breakout the stats for each card. I'd like to see the following "improvements":

- GPU/x is useless if it doesn't tell us which PCIe slot it's in... maybe add a new switch to "expert mode" so that it doesn't say GPU/0 .... GPU/1 .... GPU/n .....  Instead, show me the PCIe device address.

- Building on the "expert mode" idea, I'd like to see the entire output string for ethminer events to be trimmed down to abbreviated or removed strings. For example, "Speed", "Mh/s", "A,R,F" & "Time:" is unnecessary if the expert user already knows what is being shown.

- Keep track of which GPU is a contributor or poor performer. Hashrate is great but I'd like to see a second line of text output after a share is accepted/rejected/failed. Basically, a second line that shows this info PER GPU instead of the hashrate... then continue on showing the hashrate until the next event.

That is all for now. :)

**Comment by AndreaLanfranchi** [comment]
> GPU/x is useless if it doesn't tell us which PCIe slot it's in... maybe add a new switch to "expert mode" so that it doesn't say GPU/0 .... GPU/1 .... GPU/n ..... Instead, show me the PCIe device address.

Simply set this before starting ethminer to have the GPU's numbered in the exact order they're seen on the BUS.

```
export CUDA_DEVICE_ORDER=PCI_BUS_ID
```

**Comment by 4a48** [comment]
@AndreaLanfranchi 
Simply ordering the PCI_BUS_ID doesn't really address the issue here. Let me explain...
Let's say you have the ASUS B250 Mining Expert motherboard in your mining rig. You have 8 GPUs installed which are identical make/model/size. As is the case with most rigs, PCIe risers are used/required. Everything appears to be working correctly but upon running ethminer you discover that GPU4 is reporting 0mh/s. Even if you ordered your GPUs by BUS_ID, does the average user really know how the ordering works? Do I trust ethminer will respect the Bus/Device order? What happens if I have both AMD and nVidia GPUs in the same rig? Does ethminer still respect the PCIe bus and order the GPU numbering in ascending order?
Adding the ability to show the PCIe device ID instead of the GPU# makes troubleshooting dead simple. No questions are left if the miner is trying to nail down with PCIe slot (or riser) is causing problems.

Am I better off forking my own version or is the Ethminer developers responsive to improvements?

**Comment by jean-m-cyr** [comment]
Forking is always an option, though not always a good one. I believe that we recently added the Pcie id to the --list-devices output for CUDA GPUs. Is this the id you're looking for?

**Comment by 4a48** [comment]
I'd like to avoid forking :)
--list-devices output shows platformID, deviceID and deviceName -- not exactly sure where the deviceID getting pulled from but all it's showing when I test it on a rig is [0] [n] deviceName (n = 0...7 -- the rig I tested that switch on has 8 GPUs)
What I'd like to see is the bus number, device number and function...
Try running:
lspci -nn | grep VGA
You'll get a list of the VGA devices on your system. For example:
08:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Device [1002:67df] (rev e7)
Field 1 : 08:00.0      : bus number (08), device number (00) and function (0)
Field 2 : 0200          : device class
Field 3 : 1002:67df  : vender ID (1002) and device ID (67df)

To be clear, I'm interested in the first field. I can easily figure out the rest once I know which bus/device number the offended device is using. I suppose you could do a big data experiment and return the hashrates for vender/device IDs to the cloud and roll it all up into a pretty report showing the hashrates for different devices are getting while mining X. One report with the average per GPU every 24 hours of uninterrupted mining would be enough.

**Comment by 4a48** [comment]
Decoding device class, vender ID and device ID is pretty easy. There is a regularly updated PCI ID Databae at:
http://pci-ids.ucw.cz

...or on their github page:
https://github.com/pciutils/pciids/blob/master/pci.ids

**Comment by jean-m-cyr** [comment]
@4a48 What I meant, why not contribute the work here? You'd have to do it in your private fork anyway.

**Comment by 4a48** [comment]
@jean-m-cyr I most likely will. Just curious - is anyone looking at ncurses or something else to improve the output presentation?

**Comment by jean-m-cyr** [comment]
@4a48 Not that I know

**Comment by 4a48** [comment]
@ Okay. Leave it with me :)

**Comment by tsauce** [comment]
If at all possible, why not just take what works, if you can replicate claymore console output, you'd at least standardize and be a step in the right direction no?

**Comment by jean-m-cyr** [comment]
Nooo....

**Comment by tsauce** [comment]
LOL Open to suggestions. I'm not saying the claymore one is top-notch... everything can be improved.

**Comment by jean-m-cyr** [comment]
Well, we've done a lot to clean that up. For example, providing consistent output between cuda and opencl GPUs.

**Comment by AndreaLanfranchi** [comment]
Aside from this

>console averaged speed output over several (3, 5, or configurable) seconds, so that the rate doesn't jump up and down and scroll down too quickly

all has been done.
